jinterface: Support custom distribution protocols without epmd

JeromeDeBretagne opened the pull request:
Here is a change I've been working on for quite some time, to expand the Jinterface compatibility to a wider set of  distribution protocols (for instance based on Unix Domain Sockets) and to remove the mandatory dependency on epmd when using a custom transport factory.

The existing Jinterface classes and interfaces have been designed originally for transport protocols using a socket port number and didn't support other protocols working without epmd. I've tried a few approaches, some of them requiring quite a lot of code duplication so I've dropped those. The trickiest part has been to keep a full backwards compatibility with existing custom factories so it was not possible to extend existing interfaces with new abstract methods.

I've settled with the following approach, by creating a new base `OtpGenericTransportFactory` abstract class which does not expect a socket port number but use the node name instead as a generic identifier. New factories can be created as subclasses of `OtpGenericTransportFactory`, I've created a working example based on Unix Domain Sockets for Android here :
https://github.com/JeromeDeBretagne/erlanglauncher/pull/9/commits/dee25b20d21bbbf269d8bd6b9da1b54a58ecc110

Finally, I've just had to adapt 4 classes: AbstractConnection, AbstractNode, OtpNode and OtpSelf to accept factory instances of the new `OtpGenericTransportFactory` in parallel of the existing socket-based implementation, the changes are quite limited in each. With such transport factories, epmd is not used at all whereas the behaviors remain totally unchanged for regular factories.

I have an end-to-end implementation working on Android based on Unix Domain Sockets and without epmd here: https://github.com/JeromeDeBretagne/erlanglauncher/pull/9 with a connection between an Erlang node using the [erl_uds_dist.erl](https://github.com/JeromeDeBretagne/otp/blob/master/lib/kernel/examples/erl_uds_dist/src/erl_uds_dist.erl) alternative distribution protocol and a Jinterface node.

Let me know what you think, if you detect any issues or have suggestions for improvements.

Thanks a lot

(Edit : update the link to the working example following the renaming from `OtpNamedTransportFactory` to `OtpGenericTransportFactory`)

JeromeDeBretagne commented:
I don't know why the builds failed due to non US-ASCII characters, that's a first for me when using UTF-8: 

```
/buildroot/otp/lib/jinterface/java_src/com/ericsson/otp/erlang/OtpNamedTransportFactory.java:4: error: unmappable character (0xC3) for encoding US-ASCII
 * Copyright 2021 J??r??me de Bretagne
                   ^
```

Am I missing something?

JeromeDeBretagne commented:
Mark OtpNamedTransportFactory with %ExternalCopyright% as was done with erl_uds_dist

JeromeDeBretagne commented:
> I don't know why the builds failed due to non US-ASCII characters, that's a first for me when using UTF-8:

Thanks @sverker for the fix in e4c3fcf , I've sent a fake force-push to trigger a new run and the builds now pass!

sverker commented:
I pushed two commits to your branch. The UTF-8 characters still caused problems on some of our test machines, so I added explicit `-encoding UTF-8` for javac.

JeromeDeBretagne commented:
> I pushed two commits to your branch. The UTF-8 characters still caused problems on some of our test machines, so I added explicit `-encoding UTF-8` for javac.

Seems like a good way to avoid this issue indeed.

Good catch for the missing entry in java_files, I had totally missed it.

JeromeDeBretagne commented:
Hi @sverker this proposal has been under testing for 6 months, do you have an update ?

JeromeDeBretagne commented:
Sure, I'll wait for your first review, then add the corresponding doc and a few tests if this is an acceptable PR

JeromeDeBretagne commented:
I've rebased with the latest master branch, squashed some commits together and updated the copyright header dates to 2022.

Let me know @sverker if you can provide a first review pass in the coming days, so that I can have enought time before the OTP 25 freeze to take your comments into account. Thanks a lot.

sverker commented:
I have some time now to take a look.

JeromeDeBretagne commented:
Hi @sverker,

As requested, I have updated the documentation in the Jinterface User's Guide. Finally, I don't see which test I could add since this change is only adding a new abstract class.

I'm crossing my fingers that it can still make it into OTP 25.0-rc2 , the implementation has been stable and hasn't changed since my initial contribution last May.

Thanks,
Jérôme

sverker commented on lib/jinterface/java_src/com/ericsson/otp/erlang/OtpNamedTransportFactory.java:
```diff
@@ -0,0 +1,87 @@
+     *
+     * @throws IOException
+     */
+    public abstract OtpTransport createTransport(final String node)
+            throws IOException;
+
+    /**
+     * Create an instance of a server-side {@link OtpServerTransport}
+     *
+     * @param node
+     *            the node name identifying the server-side transport to create
+     *
+     * @return a new transport object
+     *
+     * @throws IOException
+     */
+    public abstract OtpServerTransport createServerTransport(final String node)
+            throws IOException;
```
@JeromeDeBretagne Did you miss my comments about possibilities to improve the new interface.

JeromeDeBretagne commented:
Hi @sverker ,

It seems there is a permission issue or I am not using the GitHub review system correctly, but my earlier feedback doesn't seem visible to anyone else but me, for a reason I don't understand yet (I can see it when logged in as "Pending"). I have PMed you about it by email.

In the meantime, my feedback should be visible to all as a new conversation instead so here is what I had answered back about 10 days ago. 

Let me know your thoughts. Thanks!
 

>     * Java 8 has the concept of interface default methods.

Using Jinterface only requires Java 6 / JDK 1.6.0 so far. I've seen that the Erlang/OTP team is always very careful when breaking compatibility; using this Java 8 feature would require to increase the Java minimum requirement. I didn't think this PR would be a good-enough reason to bump the whole Jinterface requirement from Java 6 to Java 8, potentially impacting customers or internal Ericsson code.

>     Could that be used to as a nicer way to extend the OtpTransportFactory interface.

Anyway, using Java 8 default methods wouldn't fix the main "issue", that the OtpTransportFactory interface makes very strong assumptions in its original design, supporting only transport protocols based on IP addresses or host names and using a mandatory socket port number (with its 3 abstract methods and the way they are used in the other classes).

Within this PR, I find it cleaner to have a new "base" class hierarchy for the EPMD-less scenario, more generic (= supporting a wide range of protocols) and quite concise with only 2 methods to implement for subclasses of OtpNamedTransportFactory. You can see a concrete sample implementation in [OtpLocalSocketTransportFactory here](https://github.com/JeromeDeBretagne/erlanglauncher/blob/881aa1728c569ae4d61552593555a51111194544/app/src/main/java/com/ericsson/otp/erlang/OtpLocalSocketTransportFactory.java)

>     * What if the EPMD publish/unpublish was abstracted through the OtpTransportFactory interface.

Same feedback, this wouldn't fix the original OtpTransportFactory interface design (with mandatory methods supporting only transport protocols based on IP addresses or host names and using a mandatory socket port number).

For this change, the OtpTransportFactory design was the main limitation I have tried to unlock in a new and more generic way. Disabling EPMD publish/unpublish is just a side effect of not relying on socket port numbers anymore. But maybe I'm missing the point you are trying to make?

>     * Is it really necessary to pass the node name when the transport object is created. Could it not be passed when constructing the factory object? If so, that would remove the need for the new createTransport methods.

If I remember correctly, yes it is. As I shared in the preambule of this PR, the difficulty has really been to plug into the existing code of Jinterface classes and interfaces. At first, I had tried to take more freedom in my design but then I realized step-by-step that I had to stay as close as possible to the existing class hiearchy.

This was more than 6 months ago (so this may not be the only constraint I faced at the time) but at least I had to remain compatible with this existing OtpNode method:
```
     * @param node
     *            the name of this node.
     *
     * @param transportFactory
     *            the transport factory to use when creating connections.
     *
     * @exception IOException
     *                if communication could not be initialized.
     *
     */
    public OtpNode(final String node,
            final OtpTransportFactory transportFactory) throws IOException {
```
which takes as arguments **the node name separately from the factory object**. This code then calls init(), then Acceptor() and finally createServerTransport() so I had to implement a similar `createServerTransport` taking the node name to integrate properly into the existing OtpNode class at least.

As a benefit for existing users of OtpTransportFactory, understanding OtpNamedTransportFactory is trivial because it reuses exactly the same 2 methods `createTransport` and `createServerTransport`.

If I could have started with a clean sheet, I would certainly have come up with a different design. For this PR, the priority has been to integrate OtpNamedTransportFactory nicely with the existing Jinterface implementation.

JeromeDeBretagne commented:
@sverker all my comments and feedback were stuck in review mode for days 😭 I believe I have found how to unlock this situation, they should be visible to all now. 

Sorry about the confusion, let me know what you think about my comments. Thanks, Jérôme

sverker commented:
My questions about default interface methods etc, all came from my observation that your extension of the OtpTransportFactory seemed a bit awkward to me. Extending OtpTransportFactory in name only. I was just probing if it was a better way to do it. Maybe not.
A more important question is if the new OtpNamedTransportFactory is generic and extendable enough. As it's an abstract class and not an interface, we can extend it in the future with new callback methods with default implementations (without demanding Java 8), so that's good.
However, to make it more generic and not assume that the node name is the only thing createTransport and createServerTransport need, how about change

createTransport(peer.node()) to createTransport(peer)
and
createServerTransport(local.node()) to createServerTransport(local)

that way the implementations can fetch whatever information they need from OtpPeer and OtpLocalNode.


About testing. We do want regression tests for new features. The way to test a new interface like this is to do a test implement of that interface and make sure the calls to it works as they should. Can you use your OtpLocalSocketTransportFactory or a simplified version of it maybe?

JeromeDeBretagne commented:
> A more important question is if the new OtpNamedTransportFactory is generic and extendable enough.

That's a good point to discuss indeed.

> to make it more generic and not assume that the node name is the only thing createTransport and createServerTransport need, how about change
> 
> createTransport(peer.node()) to createTransport(peer) and createServerTransport(local.node()) to createServerTransport(local)
> 
> that way the implementations can fetch whatever information they need from OtpPeer and OtpLocalNode.

I've applied your suggested changes and I haven't found any blocking points when switching to this more generic implementation, so it's looking good :smile: 

Let me know if this is exactly what you had in mind, based on the last commit "Switch to a more generic implementation" I've just added.
 
Here is a follow-up question now : how should I rename OtpNamedTransportFactory since it's not using node names anymore?
OtpGenericTransportFactory?
OtpExtendedTransportFactory?
Any other suggestions?

> About testing. [...] The way to test a new interface like this is to do a test implement of that interface and make sure the calls to it works as they should. Can you use your OtpLocalSocketTransportFactory or a simplified version of it maybe?

I wish I could reuse my code but its implementation relies on Android-specific classes that are not part of OpenJDK (to use Unix Domain Sockets as an alternative protocol) :disappointed:  Unix Domain Sockets were only added in Java 16 in OpenJDK so I can't use those new classes either in OTP. I fully understand the value to add tests but I don't see an easy way without implementing a full distribution protocol in Java in this case, that would be a huge undertaking.

sverker commented on lib/jinterface/java_src/com/ericsson/otp/erlang/AbstractConnection.java:
```diff
@@ -1098,12 +1098,12 @@ protected void doConnect() throws IOException, OtpAuthException {
-                System.out.println("-> MD5 CONNECT TO " + node);
+                System.out.println("-> MD5 CONNECT TO " + peer.node());
             }
             doConnect();
```
The `peer` argument is redundant as the we also have a member variable named `peer` which is the same thing.
I also suggest rename the three doConnect functions, something like this:

doConnect(OtpPeer peer) -> doGenericConnect()
doConnect(int port) -> doPortConnect(int port)
doConnect() -> handshake_we_started()        // this is what the corresponding Erlang function is called in dist_util.erl

JeromeDeBretagne replied:
I have switched the new class to OtpGenericTransportFactory as discussed and I have renamed the three doConnect functions with three different names.

Let me know if you have any other suggestions, then I will squash all the code changes into one commit and the documentation into another one.

I still have to work on adding a test, hopefully in the coming days.

sverker reviewed (commented):
Yes, this is what I had in mind.

OtpGenericTransportFactory sounds ok.

sverker commented:
An "easy" way out for testing could be to implement the existing epmd-socket-based transport using the OtpGenericTransportFactory interface.

JeromeDeBretagne commented:
> An "easy" way out for testing could be to implement the existing epmd-socket-based transport using the OtpGenericTransportFactory interface.

Thanks for this suggestion, I will investigate it.

I also have another idea, it may be smaller and simpler just for testing (TBC) : creating an Epmd-less socket distribution based on predefined socket port numbers.

JeromeDeBretagne commented:
> > creating an Epmd-less socket distribution based on predefined socket port numbers.
> 
> Sure, whatever is easiest. The actual transport protocol used is not important.

I've added a test case checking the connection between 2 nodes using a minimalist test factory based on OtpGenericTransportFactory.

Let me know what you think about it, @sverker, and if there is anything else missing to consider merging this new feature.

JeromeDeBretagne commented on lib/jinterface/test/jinterface_SUITE.erl:
```diff
@@ -114,14 +114,15 @@ end_per_group(_GroupName, Config) ->
+     maps,                      % Maps.java
+     nodename,                  % Nodename.java
+     register_and_whereis,      % RegisterAndWhereis.java
+     transport_factory          % TransportFactoryTest.java
```
Done, I have reverted this part to keep the unsorted mess :smile:

JeromeDeBretagne commented on lib/jinterface/test/jinterface_SUITE_data/Makefile.src:
```diff
@@ -36,19 +36,20 @@ JINTERFACE_CLASSPATH = @jinterface_classpath@
-	FunEquals.java \
-	CoreMatchBind.java
+	RegisterAndWhereis.java \
+	TransportFactoryTest.java
```
Done, reverted here too.

sverker commented:
Pushed a commit that adds` -encoding UTF-8` to test Makefile's as well.

sverker commented:
@JeromeDeBretagne The tests are happy, I'm happy. I suggest you squash implementation, tests and docs into one commit. Maybe keep the `-enconding UTF-8` fixes into a preceding commit. Then we are go for merge.

JeromeDeBretagne commented:
> @JeromeDeBretagne The tests are happy, I'm happy. I suggest you squash implementation, tests and docs into one commit. Maybe keep the `-enconding UTF-8` fixes into a preceding commit. Then we are go for merge.

Done, commits squashed together as suggested. I'm really happy too to see this new feature being merged :smiley: Thanks again @sverker for your review

sverker commented:
Merged! Thanks for the contribution and your patience.
